### PR TITLE
List box, Combo box, Dropdown menus, Context menu

### PR DESCRIPTION
**Pull request by h00k00**
Fixes https://github.com/vaadin/vaadin-material-theme/issues/7

Latest versions of components also taken into use

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-material-theme/17)

### CODE REVIEW

**Comment by kylefarris** [comment]
This would be awesome to have right now. Thanks for submitting this PR? I know its that time of year but can we get this merged in?

**Comment by jouni** [comment]
@kylefarris, we’ll do a proper review of this PR during this and next week, and merge it after that. We’ll probably try to make a new release also at that point.

**Comment by platosha** [comment]
I think it would make sense to adjust the list-box, item and the overlay styles first, as their look affects the other elements.

Expected:
- The overlay has subtly rounded corners.  The item’s corners are not rounded.
- Menu overlay has offsets of `8dp` from top and bottom, `0` offset from left and right edges
- Some padding from the left of the item.
screen shot 2018-01-04 at 16 00 00
screen shot 2018-01-04 at 15 58 35

Actual, as seen in `<vaadin-context-menu>` demo:
- The overlay’s corners are sharp, but the item corners are rounded.
- The overlay has a `4px` offset from all edges.
- Padding on the left of the item is too small.
screen shot 2018-01-04 at 15 56 58

**Comment by platosha** [comment]
Expected: the menu overlay has a dense shadow of `2dp`:
screen shot 2018-01-04 at 16 15 10

Actual: the menu overlay shadow is too transparent and blurry:
screen shot 2018-01-04 at 16 15 10

The Guidelines seem to use `2dp` shadow for menus, despite not naming this value explicitly. Both the Menu from the Material Components for the Web and `<paper-dropdown-menu>` use the same `2dp` shadow too.

The correct shadow value corresponding to `2dp` could be copied from [`<paper-styles>`](https://github.com/PolymerElements/paper-styles/blob/739c14c9114dd3f80172e43d4ea2c5faf9bae132/shadow.html#L27-L31).

**Comment by platosha** [comment]
Here are some easily noticeable issues of material `<vaadin-dropdown-menu>`.

Expected: prefilled value does not have offset from the left. As seen in the [Text Fields styles in the Guidelines](https://material.io/guidelines/components/text-fields.html#text-fields-states):
screen shot 2018-01-04 at 16 57 10

Actual: dropdown’s value has offset from the left, strangely different across browsers.
In Chrome:
screen shot 2018-01-04 at 16 59 52

In Firefox:
screen shot 2018-01-04 at 16 58 55

<hr>

Expected: single underline below the value box.

Actual: there is one more underline between the label and the value box on some states, namely, error state, active state for read-only menu. Try to focus and blur these elements in the demo to reproduce.
screen shot 2018-01-04 at 17 06 02
screen shot 2018-01-04 at 17 06 06

<hr>

Expected: focused dropdown menu looks like in the Guidelines for Text Field: underline uses primary color.

Actual: when focused with mouse, the underline remains black:
screen shot 2018-01-04 at 17 11 02

When focused with a keyboard <kbd>Tab</kbd> key, there are double underlines and an extra outline displayed
screen shot 2018-01-04 at 17 12 37

<hr>

Expected: required asterisk indicator is next to the label and uses the same label color:
screen shot 2018-01-04 at 17 15 58

Actual: required asterisk indicator has random position somewhere to the right and uses a different color on some cases:
screen shot 2018-01-04 at 17 17 01
screen shot 2018-01-04 at 17 17 09

**Comment by platosha** [comment]
In Firefox with `<vaadin-dropdown-menu>` dark theme, on mouse hover, the label colour changes to unreadable dark black:
screen shot 2018-01-04 at 17 23 13
screen shot 2018-01-04 at 17 21 58
screen shot 2018-01-04 at 17 22 03

**Comment by platosha** [comment]
Expected: “Label” of the empty combo-box is displayed in the same place, where the value “Option 1” of non-empty combo-box is.

Actual: “Label” is visually higher than “Option 1”.

screen shot 2018-01-11 at 14 37 15

---

Open a combo-box and navigate the options list using keyboard <kbd>↑</kbd> and <kbd>↓</kbd> keys.

Expected: currently “focused”[1] item is highlighted in the list. Light grey background is used for the keyboard highlight (same as for mouse hover).

Actual: currently “focused” item is styled exactly the same as the regular items, and thus it’s impossible to tell which item is “focused” in the list.
 
combo-box-keyboard-item

---

Expected: when field is focused or blurred with keyboard <kbd>Tab</kbd> (<kbd>Shift</kbd>+<kbd>Tab</kbd>), the underline changes color and may ripple.

Actual: there is a weird ripple overline animation between the label and the value box when focused or blurred from a keyboard.

combo-box-focus-overline

File: mixins/menu-overlay.html [hunk 1]
```diff
@@ -0,0 +1,62 @@
+<link rel="import" href="../color.html">
+<link rel="import" href="overlay.html">
+
+<dom-module id="material-menu-overlay" theme-for="vaadin-dropdown-menu-overlay vaadin-combo-box-overlay vaadin-context-menu-overlay">
```
Comment (jouni) [inline mixins/menu-overlay.html]: Should not specify `theme-for` for this module. The individual modules should opt-in to include this module (like they actually are right now, which probably causes them to be loaded twice).

**Comment by web-padawan** [comment]
Reviewed 2 of 27 files at r1, 1 of 16 files at r5, 1 of 3 files at r7.
Review status: 4 of 31 files reviewed at latest revision, 3 unresolved discussions.

---

*[mixins/menu-overlay.html, line 37 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17#-L4AbpT8Yw2w6aNezu3o:-L4AbpT9Yk4PpcP9hjC8:b-wd97ti) ([raw file](https://github.com/vaadin/vaadin-material-theme/blob/d0d1fc382b55da036140bc1d5426f499a2cf2b5e/mixins/menu-overlay.html#L37)):*
> ```HTML
> 
>       vaadin-list-box [tabindex] {
>         min-height: 36px;
> ```

should be `48px` by default per MD [specs](https://material.io/guidelines/components/menus.html#menus-specs) for menus

---

*[mixins/menu-overlay.html, line 38 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17#-L4AcPwCoVJsQioxmSsn:-L4AcPwCoVJsQioxmSso:bset7fx) ([raw file](https://github.com/vaadin/vaadin-material-theme/blob/d0d1fc382b55da036140bc1d5426f499a2cf2b5e/mixins/menu-overlay.html#L38)):*
> ```HTML
>       vaadin-list-box [tabindex] {
>         min-height: 36px;
>         padding: 4px 8px;
> ```

should be `16px` at the left side per MD [specs](https://material.io/guidelines/components/menus.html#menus-specs) for menus

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17)*

**Comment by web-padawan** [comment]
Reviewed 1 of 16 files at r5.
Review status: 5 of 31 files reviewed at latest revision, 4 unresolved discussions.

---

*[bower.json, line 56 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17#-L4AokcM8yfcnH9lo-lA:-L4AokcM8yfcnH9lo-lB:b-5neiz9) ([raw file](https://github.com/vaadin/vaadin-material-theme/blob/d0d1fc382b55da036140bc1d5426f499a2cf2b5e/bower.json#L56)):*
> ```JSON
>     "iron-icons": "polymerelements/iron-icons#^2.0.1"
>   },
>   "resolutions": {
> ```

there is a bunch of extra resolutions needed to run `bower install` from the current PR:
```
    "vaadin-button": "master",
    "vaadin-checkbox": "master",
    "vaadin-item": "master",
    "vaadin-text-field": "master",
    "vaadin-icons": "master"
```

---

*[mixins/menu-overlay.html, line 37 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17#-L4AbpT8Yw2w6aNezu3o:-L4AzjTzYKhkh-SUa7ZC:bquh122) ([raw file](https://github.com/vaadin/vaadin-material-theme/blob/d0d1fc382b55da036140bc1d5426f499a2cf2b5e/mixins/menu-overlay.html#L37)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

should be `48px` by default per MD [specs](https://material.io/guidelines/components/menus.html#menus-specs) for menus
</blockquote></details>

upd: my bad, that was for mobile, and for desktop the `32px`

---

*[mixins/menu-overlay.html, line 38 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17#-L4AcPwCoVJsQioxmSsn:-L4AztKCAg26siq0A3YM:bqtyifr) ([raw file](https://github.com/vaadin/vaadin-material-theme/blob/d0d1fc382b55da036140bc1d5426f499a2cf2b5e/mixins/menu-overlay.html#L38)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

should be `16px` at the left side per MD [specs](https://material.io/guidelines/components/menus.html#menus-specs) for menus
</blockquote></details>

upd: my bad, that was for mobile, and for desktop the `24px`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-material-theme/17)*

**Pull request merged**